plugins/outputs/sql/sql.go

```diff
@@ -200,28 +213,130 @@ func (p *SQL) generateInsert(tablename string, columns []string) string {
 		strings.Join(placeholders, ","))
 }
 
-func (p *SQL) tableExists(tableName string) bool {
-	stmt := strings.ReplaceAll(p.TableExistsTemplate, "{TABLE}", quoteIdent(tableName))
+func (p *SQL) fetchColumns(tableName string) error {
+	stmt := p.ColumnQueryTemplate
+	stmt = strings.ReplaceAll(stmt, "{TABLE}", quoteStr(tableName))
+
+	columns, err := p.db.Query(stmt)
+
+	if err != nil {
+		return err
+	}
+
+	defer columns.Close()
+
+	for columns.Next() {
+		var columnName string
+
+		if err := columns.Scan(&columnName); err != nil {
+			return err
+		}
+
+		if !slices.Contains(p.tables[tableName], columnName) {
+			p.tables[tableName] = append(p.tables[tableName], columnName)
+		}
+	}
+
+	return nil
+}
+
+func (p *SQL) createTableIfDoesNotExist(metric telegraf.Metric) error {
+	createStmt := p.generateCreateTable(metric)
+	_, err := p.db.Exec(createStmt)
+	if err != nil {
+		return err
+	}
+
+	err = p.fetchColumns(metric.Name())
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (p *SQL) addColumnIfDoesNotExist(tablename, column, columnType string) error {
+	if !slices.Contains(p.tables[tablename], column) {
+		createColumn := p.generateAddColumn(tablename, column, columnType)
+
+		_, err := p.db.Exec(createColumn)
+
+		if err != nil {
+			return err
+		}
+
+		err = p.fetchColumns(tablename)
+
+		if err != nil {
+			return err
+		}
```

Comment:
Why do you fetch the columns again?

Reply:
fetchColumns populates the cache, we could update it directly, but that way we would duplicate code.

Reply:
Yeah what I meant is that in the previous version you also fetched the column when updating the table...